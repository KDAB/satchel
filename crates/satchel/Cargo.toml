[package]
name = "satchel"
version = "0.1.1"
license = "MIT OR Apache-2.0"
repository = "https://github.com/KDAB/satchel/"
edition = "2024"
description = "Test registration and discovery crate. Satchel uses linkme for test registration and exposes APIs for test discovery."

[dependencies]
libtest-mimic = "0.8"
linkme = "0.3"
satchel-macro = "0.1.0"

