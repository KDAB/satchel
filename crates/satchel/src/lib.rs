#![no_std]

pub use satchel_macro::{bench, test};

pub type TestFn = fn();

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum TestKind {
    Unit,
    Benchmark,
}

#[derive(Debug, Clone)]
pub struct ShouldPanic {
    pub expected: Option<&'static str>,
}

#[derive(Debug, Clone)]
pub struct Ignore {
    pub reason: Option<&'static str>,
}

#[derive(Debug, Clone)]
pub struct TestCase {
    pub name: &'static str,
    pub module_path: &'static str,
    pub kind: TestKind,
    pub test_fn: TestFn,
    pub should_panic: Option<ShouldPanic>,
    pub ignore: Option<Ignore>,
}

pub mod test_harness {
    pub use crate::{Ignore, ShouldPanic, TestCase};
    use linkme::distributed_slice;

    #[doc(hidden)]
    #[distributed_slice]
    pub static TESTS: [TestCase];
}

#[doc(hidden)]
pub fn extract_crate_name(module_path: &str) -> &str {
    module_path
        .split("::")
        .next()
        .expect("Split never returns a empty iterator")
}

#[doc(hidden)]
pub fn get_tests_for_crate(crate_prefix: &str) -> impl Iterator<Item = &'static TestCase> {
    let crate_name = extract_crate_name(crate_prefix);
    test_harness::TESTS
        .iter()
        .filter(move |case| case.module_path.starts_with(crate_name))
}

#[macro_export]
macro_rules! get_tests {
    () => {
        ::satchel::get_tests_for_crate(::core::module_path!())
    };
}

#[cfg(test)]
mod tests {
    use super::extract_crate_name;

    #[test]
    fn handles_empty_string() {
        assert_eq!(extract_crate_name(""), "");
    }

    #[test]
    fn extracts_simple_crate_name() {
        assert_eq!(extract_crate_name("mycrate"), "mycrate");
    }

    #[test]
    fn extracts_crate_name_with_module() {
        assert_eq!(extract_crate_name("mycrate::foo"), "mycrate");
        assert_eq!(extract_crate_name("mycrate::foo::bar"), "mycrate");
    }
}
