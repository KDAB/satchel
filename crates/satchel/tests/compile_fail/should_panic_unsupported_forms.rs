mod common;
use satchel::test;

// Multiple positional arguments
#[test]
#[should_panic("a", "b")]
fn multi_positional_is_invalid() {}

// Mixed positional and named arguments (named first)
#[test]
#[should_panic(expected = "boom", "other")]
fn mixed_positional() {}

// Mixed positional and named arguments (positional first)
#[test]
#[should_panic("boom", expected = "boom")]
fn mixed_reverse() {}

// Unknown key in named arguments
#[test]
#[should_panic(bad = "format")]
fn bad_should_panic_form() {}

// Non-string value for named argument
#[test]
#[should_panic(message = 123)]
fn should_panic_non_string() {}

// Duplicate expected arguments
#[test]
#[should_panic(expected = "a", expected = "b")]
fn dup_expected() {}

// Duplicate should_panic attributes
#[test]
#[should_panic]
#[should_panic(expected = "boom")]
fn duplicate_should_panic() {}

fn main() {}