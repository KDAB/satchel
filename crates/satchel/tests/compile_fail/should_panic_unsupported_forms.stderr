error: unsupported form in #[should_panic]; allowed: #[should_panic], #[should_panic(expected = "..."), #[should_panic = "..."], #[should_panic("...")]
 --> tests/compile_fail/should_panic_unsupported_forms.rs:6:1
  |
6 | #[should_panic("a", "b")]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^

error: unsupported form in #[should_panic]; allowed: #[should_panic], #[should_panic(expected = "..."), #[should_panic = "..."], #[should_panic("...")]
  --> tests/compile_fail/should_panic_unsupported_forms.rs:11:1
   |
11 | #[should_panic(expected = "boom", "other")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unsupported form in #[should_panic]; allowed: #[should_panic], #[should_panic(expected = "..."), #[should_panic = "..."], #[should_panic("...")]
  --> tests/compile_fail/should_panic_unsupported_forms.rs:16:1
   |
16 | #[should_panic("boom", expected = "boom")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unsupported form in #[should_panic]; allowed: #[should_panic], #[should_panic(expected = "..."), #[should_panic = "..."], #[should_panic("...")]
  --> tests/compile_fail/should_panic_unsupported_forms.rs:21:16
   |
21 | #[should_panic(bad = "format")]
   |                ^^^^^^^^^^^^^^

error: unsupported form in #[should_panic]; allowed: #[should_panic], #[should_panic(expected = "..."), #[should_panic = "..."], #[should_panic("...")]
  --> tests/compile_fail/should_panic_unsupported_forms.rs:26:16
   |
26 | #[should_panic(message = 123)]
   |                ^^^^^^^^^^^^^

error: duplicate #[should_panic] expected message
  --> tests/compile_fail/should_panic_unsupported_forms.rs:31:32
   |
31 | #[should_panic(expected = "a", expected = "b")]
   |                                ^^^^^^^^

error: duplicate #[should_panic] attribute
  --> tests/compile_fail/should_panic_unsupported_forms.rs:37:1
   |
37 | #[should_panic(expected = "boom")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
